### PR TITLE
speedup attribute style access and tab completion

### LINKED ISSUE #4741: Attribute style access is slow
I appreciate xarray's ability to use attribute style access ``ds.foo`` as an alternative to ``ds["foo"]`` as it requires less characters/keystrokes and has less 'visual clutter'.

A drawback is that it can be much slower as lookup time seems to display `O(n)` behaviour instead of `O(1)` with `n` being the number of variables in the dataset. For e.g. `n=100` it is approximately 100 times slower than dictionary-style access:

    # Dataset with many (100) variables
    ds = xr.Dataset({f'var{v}': [] for v in range(100)})

    %timeit ds['var0']
    462 µs ± 1.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

    %timeit ds.var0
    47.1 ms ± 205 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

``dir()`` and ``_ipython_key_completions_()`` which are used for e.g. tab completion in iPython are equally slow:

    %timeit dir(ds)
    47 ms ± 163 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

    %timeit ds._ipython_key_completions_()
    46.8 ms ± 210 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

I would like to see xarray having much better performance for attribute style access.

### PR DESCRIPTION
**Pull request by rhkleijn**
- [x] Closes #4741 
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

- make _attr_sources and _item_sources lazy
- make lookup of virtual dimension coordinates lazy
- make helper class more general so it can be used for lookup of both virtual dimension and level coordinates

### CODE REVIEW

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -1340,19 +1340,22 @@ def __deepcopy__(self, memo=None) -> "Dataset":
...
-        ]
+    def _item_sources(self) -> Iterable[Mapping[Hashable, Any]]:
+        """Places to look-up items for key-completion"""
+        yield self.data_vars
+        yield self.coords
+
+        # virtual coordinates
+        yield HybridMappingProxy(keys=self.dims, mapping=self)
+
+        # uses empty {} -- everything here can already be found in self.coords.
```
Comment (mathause) [inline xarray/core/dataset.py:1357]:
```suggestion
        # use empty dict -- everything here can already be found in self.coords.
```

**Review by mathause — APPROVED** [review]
LGTM (as far as I understand it)

File: xarray/core/utils.py [hunk 2]
```diff
@@ -435,6 +435,31 @@ def FrozenDict(*args, **kwargs) -> Frozen:
     return Frozen(dict(*args, **kwargs))
 
 
+class HybridMappingProxy(Mapping[K, V]):
+    """Implements the Mapping interface. Uses the wrapped mapping for item lookup
+    and a separate wrapped keys collection for iteration.
+
+    Can be used to construct a mapping object from another dict-like object without
+    eagerly accessing its items or when a mapping object is expected but only
+    iteration over keys is actually used.
```
Comment (max-sixty) [inline xarray/core/utils.py:444]: Could we note that it's the caller's responsibility to ensure that `keys` are the keys of `mapping`, i.e. that this doesn't validate that?

**Comment by max-sixty** [comment]
Looks great, thanks @rhkleijn

File: xarray/core/dataarray.py [hunk 3]
```diff
@@ -721,18 +726,20 @@ def __delitem__(self, key: Any) -> None:
...
-    def _item_sources(self) -> List[Mapping[Hashable, Any]]:
-        """List of places to look-up items for key-completion"""
-        return [
-            self.coords,
-            {d: self.coords[d] for d in self.dims},
-            LevelCoordinatesSource(self),
-        ]
+    def _item_sources(self) -> Iterable[Mapping[Hashable, Any]]:
+        """Places to look-up items for key-completion"""
+        yield self.coords
```
Comment (dcherian) [inline xarray/core/dataarray.py:737]:
```suggestion
        yield self._coords
```

File: xarray/core/dataset.py [hunk 4]
```diff
@@ -1340,19 +1340,22 @@ def __deepcopy__(self, memo=None) -> "Dataset":
...
-        return [
-            self.data_vars,
-            self.coords,
-            {d: self[d] for d in self.dims},
-            LevelCoordinatesSource(self),
-        ]
+    def _item_sources(self) -> Iterable[Mapping[Hashable, Any]]:
+        """Places to look-up items for key-completion"""
+        yield self.data_vars
+        yield self.coords
```
Comment (dcherian) [inline xarray/core/dataset.py:1352]:
```suggestion
        yield self._coord_names
```

**Review by dcherian — COMMENTED** [review]
using `_coords` and `_coord_names` should improve performance a little more

File: xarray/core/dataset.py (hunk 1 above)
Comment (rhkleijn) [reply xarray/core/dataset.py:1357]: thanks, will do

File: xarray/core/utils.py (hunk 2 above)
Comment (rhkleijn) [reply xarray/core/utils.py:444]: That is a good suggestion. I will add a note.

File: xarray/core/dataarray.py (hunk 3 above)
Comment (rhkleijn) [reply xarray/core/dataarray.py:737]: For iterating the keys your suggestion is definitely faster. The lookup result type would however change from `DataArray` (with `self.coords`) to `IndexVariable` (with `self._coords`) and we would lose the virtual coordinates (e.g. a dim without coordinates). 
I will change this to `yield HybridMappingProxy(keys=da._coords, mapping=da.coords)`. This improves performance of iteration without changing semantics. Thanks for the suggestion!

File: xarray/core/dataset.py (hunk 4 above)
Comment (rhkleijn) [reply xarray/core/dataset.py:1352]: Thanks. Similar to the previous one I will change this one to
```yield HybridMappingProxy(keys=self._coord_names, mapping=self.coords)```

**Review by shoyer — APPROVED** [review]
nice work!

**Comment by mathause** [comment]
Thanks @rhkleijn!

**Pull request merged**